flink-runtime/src/main/java/org/apache/flink/runtime/checkpoint/PendingCheckpoint.java

```diff
@@ -191,6 +194,22 @@ void setStatsCallback(@Nullable PendingCheckpointStats trackerCallback) {
 		this.statsCallback = checkNotNull(trackerCallback);
 	}
 
+	/**
+	 * Sets the handle for the canceller to this pending checkoint.
+	 * 
+	 * @return true, if the handle was set, false, if the checkpoint is already disposed;
+	 */
+	public boolean setCancellerHandle(ScheduledFuture<?> cancellerHandle) {
+		synchronized (lock) {
+			if (!discarded) {
+				this.cancellerHandle = cancellerHandle;
```

Comment:
I know this method will only be called once in `CheckpointCoordinator`. However, in order to make this class self contained I think it would either be good to fail if we have already set an `cancellerHandle` or cancel the old
canceller to avoid an outdated discard call.